### PR TITLE
Simplify java client

### PR DESCRIPTION
**Pull request by darionyaphet**

### CODE REVIEW

File: client/src/main/java/com/vesoft/nebula/AbstractClient.java [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+import com.facebook.thrift.transport.TTransport;
+
+import com.google.common.collect.Lists;
+import com.google.common.net.HostAndPort;
+import com.google.common.net.InetAddresses;
+import com.vesoft.nebula.graph.ErrorCode;
+import java.util.List;
+import java.util.Random;
+
+public abstract class AbstractClient implements Client {
```
Comment (critical27) [inline client/src/main/java/com/vesoft/nebula/AbstractClient.java]: 👍

File: client/src/main/java/com/vesoft/nebula/client/graph/async/AsyncGraphClientImpl.java [hunk 2]
```diff
@@ -134,11 +74,12 @@ public int connect(String username, String password) {
 
             try {
                 manager = new TAsyncClientManager();
-                transport = new TNonblockingSocket(address.getHost(), address.getPort(), timeout);
+                transport = new TNonblockingSocket(address.getHostText(),
```
Comment (critical27) [inline client/src/main/java/com/vesoft/nebula/client/graph/async/AsyncGraphClientImpl.java]: Maybe move to doConnect?

File: client/src/main/java/com/vesoft/nebula/client/storage/StorageClientImpl.java [hunk 3]
```diff
@@ -213,12 +184,12 @@ public void run() {
...
 
-    private boolean doPut(int space, HostAddr leader, PutRequest request) {
-        StorageService.Client client = connect(leader);
-        if (client == null) {
+    private boolean doPut(String space, HostAddr leader, PutRequest request) {
+        if (!clientMap.containsKey(addr)) {
             return false;
         }
 
+        StorageService.Client client = clientMap.get(leader);
```
Comment (critical27) [inline client/src/main/java/com/vesoft/nebula/client/storage/StorageClientImpl.java]: Maybe later we need to put `clientMap.get(leader);` into a separate function, it seems the client can not recover if its connection is broken (at least in jepsen test).

File: client/src/main/java/com/vesoft/nebula/client/graph/async/AsyncGraphClientImpl.java (hunk 2 above)
Comment (darionyaphet) [reply client/src/main/java/com/vesoft/nebula/client/graph/async/AsyncGraphClientImpl.java]: yep. I'm still doing it.

File: client/src/main/java/com/vesoft/nebula/AsyncAbstractClient.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
+package com.vesoft.nebula;
```
Comment (wfystx) [inline client/src/main/java/com/vesoft/nebula/AsyncAbstractClient.java]: Copyright

File: client/src/main/java/com/vesoft/nebula/AbstractClient.java [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+ * This source code is licensed under Apache 2.0 License,
+ * attached with Common Clause Condition 1.0, found in the LICENSES directory.
+ */
+
+package com.vesoft.nebula;
+
+import com.facebook.thrift.TException;
+import com.facebook.thrift.protocol.TProtocol;
+import com.facebook.thrift.transport.TTransport;
+import static com.google.common.base.Preconditions.checkArgument;
```
Comment (wfystx) [inline client/src/main/java/com/vesoft/nebula/AbstractClient.java]: CheckStyle would probably remind me to move the static import to the top, just below package. not pretty sure though :(

File: client/src/main/java/com/vesoft/nebula/AbstractClient.java [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+     *
+     * @param addresses       The addresses of graph services.
+     * @param timeout         The timeout of RPC request.
+     * @param connectionRetry The number of retries when connection failure.
+     * @param executionRetry  The number of retries when execution failure.
+     */
+    public AbstractClient(List<HostAndPort> addresses, int timeout,
+                          int connectionRetry, int executionRetry) {
+        checkArgument(timeout > 0);
+        checkArgument(connectionRetry > 0);
```
Comment (wfystx) [inline client/src/main/java/com/vesoft/nebula/AbstractClient.java]: maybe also check executionRetry

File: client/src/main/java/com/vesoft/nebula/AsyncAbstractClient.java [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+    public void close() {
+        service.shutdown();
+        transport.close();
+        try {
+            manager.stop();
+        } catch (InterruptedException e) {
+            e.printStackTrace();
+        }
+    }
+}
```
Comment (wfystx) [inline client/src/main/java/com/vesoft/nebula/AsyncAbstractClient.java]: Great idea of abstract clients! 🥇

File: client/src/main/java/com/vesoft/nebula/client/graph/GraphClient.java [hunk 8]
```diff
@@ -48,6 +41,6 @@
      * @throws NGQLException       the nebula exception
      * @throws TException          the thrift exception
      */
-    public ResultSet executeQuery(String statement)
+    public abstract ResultSet executeQuery(String statement)
             throws ConnectionException, NGQLException, TException;
```
Comment (wfystx) [inline client/src/main/java/com/vesoft/nebula/client/graph/GraphClient.java]: "abstract" could be removed

**Comment by darionyaphet** [comment]
thanks guys Currently I'm still working on it :)

File: client/src/main/java/com/vesoft/nebula/AbstractClient.java (hunk 5 above)
Comment (darionyaphet) [reply client/src/main/java/com/vesoft/nebula/AbstractClient.java]: yep it's mistake

**Pull request merged**